packages/app/src/lib/workers/sqlite/worker.ts

```diff
@@ -0,0 +1,664 @@
+  #heartbeatInterval: NodeJS.Timeout | null = null;
+  #status: Partial<SqliteStatus> = {};
+  #backend: BackendInterface | null = null;
+  #ensuredProfiles = new Set<string>();
+  #eventChannel: AsyncChannel<EventBatch>;
+  #statementChannel = new AsyncChannel<StatementBatch>();
+  #resultChannel = new AsyncChannel<ApplyResultBatch>();
+  #pendingBatches = new Map<string, (result: ApplyResultBatch) => void>();
+
+  constructor() {
+    this.#workerId = crypto.randomUUID();
+    this.#eventChannel = new AsyncChannel();
+  }
+
+  initialize(ports: { backendPort: MessagePort; statusPort: MessagePort }) {
+    // Monitor port health
+    ports.backendPort.onmessageerror = (error) => {
+      console.error("SQLite worker: Backend port message error", error);
+      this.#isConnectionHealthy = false;
+    };
+
+    ports.statusPort.onmessageerror = (error) => {
+      console.error("SQLite worker: Status port message error", error);
+      this.#isConnectionHealthy = false;
+    };
+
+    this.#status = reactiveWorkerState<SqliteStatus>(ports.statusPort, true);
+
+    this.#backend = messagePortInterface<{}, BackendInterface>(
+      ports.backendPort,
+      {},
+    );
+
+    this.#status.workerId = this.#workerId;
+    this.#status.isActiveWorker = false; // Initialize to false for reactive state tracking
+    this.#status.vfsType = undefined; // Will be set after database initialization
+    console.log("SQLite Worker id", this.#workerId);
+
+    const callback = async () => {
+      console.log(
+        "Sqlite worker lock obtained: Active worker id:",
+        this.#workerId,
+      );
+      this.#status.isActiveWorker = true;
+      this.startHeartbeat();
+
+      globalThis.addEventListener("error", this.cleanup);
+      globalThis.addEventListener("unhandledrejection", this.cleanup);
+
+      try {
+        await initializeDatabase("/mini.db");
+
+        // initialise DB schema (should be idempotent)
+        console.time("initSql");
+        await this.runSavepoint({ name: "init", items: initSql });
+        console.timeEnd("initSql");
+
+        this.#status.vfsType = getVfsType() || undefined;
+        console.log("SQLite Worker using VFS:", this.#status.vfsType);
+
+        const sqliteChannel = new MessageChannel();
+        messagePortInterface<SqliteWorkerInterface, {}>(
+          sqliteChannel.port1,
+          this.getSqliteInterface(),
+        );
+        this.#backend?.setActiveSqliteWorker(sqliteChannel.port2);
+        this.materializer();
+        this.listenStatements();
+        this.listenResults();
+        await new Promise(() => {});
+      } catch (e) {
+        console.error("SQLite worker: Fatal error", e);
+        this.cleanup();
+        throw e;
+      }
+    };
+
+    navigator.locks
+      .request(
+        "sqlite-worker-lock",
+        { mode: "exclusive", signal: AbortSignal.timeout(LOCK_TIMEOUT_MS) },
+        callback,
+      )
+      .catch(async (error) => {
```

Comment:
I think we actually need to do this in an infinite loop.

When we didn't have the timeout it would just wait until it was the only worker and then it would get the lock until it was closed. With the timeout though, we need to make sure that we keep trying to get the lock, in case the existing active worker tab gets closed and this worker needs to step in for it.